db/active_replicator_common.go

```diff
@@ -24,6 +25,127 @@ type activeReplicatorCommon struct {
 	replicationStats      *BlipSyncStats
 	onReplicatorComplete  ReplicatorCompleteFunc
 	lock                  sync.RWMutex
+	ctx                   context.Context
+	ctxCancel             context.CancelFunc
+}
+
+func newActiveReplicatorCommon(config *ActiveReplicatorConfig, direction ActiveReplicatorDirection) *activeReplicatorCommon {
+
+	var replicationStats *BlipSyncStats
+	if direction == ActiveReplicatorTypePush {
+		replicationStats = BlipSyncStatsForSGRPush(config.ReplicationStatsMap)
+	} else {
+		replicationStats = BlipSyncStatsForSGRPull(config.ReplicationStatsMap)
+	}
+
+	return &activeReplicatorCommon{
+		config:           config,
+		state:            ReplicationStateStopped,
+		replicationStats: replicationStats,
+	}
+}
+
+// reconnect synchronously calls the given _connectFn until successful, or times out trying. Retry loop can be stopped by cancelling a.ctx
+func (a *activeReplicatorCommon) reconnect(_connectFn func() error) {
```

Comment:
Is there any way for users to identify that a replication is in 'connecting' state?  If not, what state is the replication going to report via _replicationStatus?

Reply:
Readded a "reconnecting" state, and also avoided the intermediate "error" state which was being set inside connect.

So state lifecycles are now:
- "Stopped" -> "Running"
- "Stopped" -> "Error" (until reconnect kicks in) -> "Reconnecting" (with lastError set on each attempt) -> "Running"